File: android/src/main/java/com/ijzerenhein/sharedelement/RNSharedElementNodeManager.java
package com.ijzerenhein.sharedelement;

import java.util.Map;
import java.util.HashMap;

import android.view.View;
import android.content.Context;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.NativeViewHierarchyManager;

class RNSharedElementNodeManager {
    private Map<Integer, RNSharedElementNode> mNodes = new HashMap<Integer, RNSharedElementNode>();
    private NativeViewHierarchyManager mNativeViewHierarchyManager;
    private Context mContext;

    void setNativeViewHierarchyManager(NativeViewHierarchyManager nativeViewHierarchyManager) {
        mNativeViewHierarchyManager = nativeViewHierarchyManager;
    }

    NativeViewHierarchyManager getNativeViewHierarchyManager() {
        return mNativeViewHierarchyManager;
    }

    void setContext(Context context) {
        mContext = context;
    }

    Context getContext() {
        return mContext;
    }

    RNSharedElementNode acquire(int reactTag, View view, boolean isParent, View ancestor, ReadableMap styleConfig) {
        synchronized (mNodes) {
            RNSharedElementNode node = mNodes.get(reactTag);
            if (node != null) {
                node.setRefCount(node.getRefCount() + 1);
                return node;
            }
            node = new RNSharedElementNode(reactTag, view, isParent, ancestor, styleConfig, mContext);
            mNodes.put(reactTag, node);
            return node;
        }
    }

    int release(RNSharedElementNode node) {
        synchronized (mNodes) {
            node.setRefCount(node.getRefCount() - 1);
            if (node.getRefCount() == 0) {
                mNodes.remove(node.getReactTag());
            }
            return node.getRefCount();
        }
    }
}

File: web/RNSharedElementNodeManager.ts
import { RNSharedElementNode } from "./RNSharedElementNode";

export class RNSharedElementNodeManager {
  private nodes = new Map<Element, RNSharedElementNode>();

  acquire(
    domNode: Element,
    isParent: boolean,
    ancestorDomNode: Element
  ): RNSharedElementNode {
    let node = this.nodes.get(domNode);
    if (node) {
      node.addRef();
      return node;
    }
    node = new RNSharedElementNode(domNode, isParent, ancestorDomNode);
    this.nodes.set(domNode, node);
    return node;
  }

  release(node: RNSharedElementNode) {
    const refCount = node.releaseRef();
    if (!refCount) {
      this.nodes.delete(node.domNode);
    }
    return refCount;
  }
}
